File: algorithms/numerical-optimization-algorithms/simulated-annealing/python/simulated_annealing.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""simulated_annealing.py

The Simulated Annealing (SA) algorithm is a probabilistic optimization technique inspired by the process of annealing in metallurgy. Here's how the algorithm works:

    Initialization: Start with an initial solution and an initial temperature.

    Iteration: Repeat the following steps until convergence or a maximum number of iterations:
        Generate a new solution by perturbing the current solution.
        Evaluate the objective function value for the new solution.
        If the new solution is better (lower objective function value), accept it as the current solution.
        If the new solution is worse, accept it with a certain probability based on the temperature and the magnitude of the deterioration.
        Decrease the temperature according to a cooling schedule.

    Termination: Return the best solution found and its corresponding objective function value.

In the example usage, we demonstrate how to minimize a simple objective function using Simulated Annealing. We define the objective function to minimize, provide an initial solution, set the initial temperature, cooling rate, and maximum number of iterations, and call the simulated_annealing function to obtain the best solution and its corresponding objective function value. Adjustments to parameters such as the temperature, cooling rate, and perturbation magnitude can be made based on the specific optimization problem being solved.

.. _PEP 0000:
    https://peps.python.org/pep-0000/
"""


from typing import Callable, Tuple
import numpy as np


def main() -> None:
    """Driving code and main function"""
    # Example: Minimize a simple objective function using Simulated Annealing
    def objective_function(x: np.ndarray) -> float:
        return np.sum(np.square(x))

    initial_solution = np.array([2.0, 3.0, 1.5])
    temperature = 100.0
    cooling_rate = 0.95
    max_iter = 1000

    best_solution, best_energy = simulated_annealing(objective_function, initial_solution, temperature, cooling_rate, max_iter)
    print("Best solution:", best_solution)
    print("Best energy:", best_energy)
 
    return None


def simulated_annealing(objective_function: Callable[[np.ndarray], float], initial_solution: np.ndarray, temperature: float, cooling_rate: float, max_iter: int) -> Tuple[np.ndarray, float]:
    """Solve optimization problems using Simulated Annealing.

    Args:
        objective_function (Callable[[np.ndarray], float]): Objective function to minimize.
        initial_solution (np.ndarray): Initial solution vector.
        temperature (float): Initial temperature parameter.
        cooling_rate (float): Cooling rate parameter.
        max_iter (int): Maximum number of iterations.

    Returns:
        Tuple[np.ndarray, float]: Best solution found and its corresponding objective function value.
    """
    current_solution = initial_solution.copy()
    current_energy = objective_function(current_solution)
    best_solution = current_solution.copy()
    best_energy = current_energy

    for _ in range(max_iter):
        proposed_solution = current_solution + np.random.normal(scale=0.1, size=current_solution.shape)
        proposed_energy = objective_function(proposed_solution)
        delta_energy = proposed_energy - current_energy

        if delta_energy < 0 or np.random.rand() < np.exp(-delta_energy / temperature):
            current_solution = proposed_solution
            current_energy = proposed_energy

            if current_energy < best_energy:
                best_solution = current_solution
                best_energy = current_energy

        temperature *= cooling_rate

    return best_solution, best_energy


if __name__ == '__main__':
    main()

File: algorithms/numerical-optimization-algorithms/simulated-annealing/java/SimulatedAnnealing.java
import java.util.Random;
import java.util.function.Function;

public class SimulatedAnnealing {

    /**
     * Solves optimization problems using Simulated Annealing.
     *
     * @param objectiveFunction Objective function to minimize.
     * @param initialSolution  Initial solution vector.
     * @param temperature      Initial temperature parameter.
     * @param coolingRate      Cooling rate parameter.
     * @param maxIter          Maximum number of iterations.
     * @return The best solution found and its corresponding objective function value as an array.
     */
    public static double[] simulatedAnnealing(Function<double[], Double> objectiveFunction, double[] initialSolution, double temperature, double coolingRate, int maxIter) {
        double[] currentSolution = initialSolution.clone();
        double currentEnergy = objectiveFunction.apply(currentSolution);
        double[] bestSolution = currentSolution.clone();
        double bestEnergy = currentEnergy;

        Random random = new Random();
        for (int iter = 0; iter < maxIter; iter++) {
            double[] proposedSolution = new double[currentSolution.length];
            for (int i = 0; i < currentSolution.length; i++) {
                proposedSolution[i] = currentSolution[i] + random.nextGaussian() * 0.1; // Perturb the current solution
            }

            double proposedEnergy = objectiveFunction.apply(proposedSolution);
            double deltaEnergy = proposedEnergy - currentEnergy;

            if (deltaEnergy < 0 || Math.random() < Math.exp(-deltaEnergy / temperature)) {
                currentSolution = proposedSolution;
                currentEnergy = proposedEnergy;

                if (currentEnergy < bestEnergy) {
                    bestSolution = currentSolution.clone();
                    bestEnergy = currentEnergy;
                }
            }

            temperature *= coolingRate; // Cool down temperature
        }

        return new double[]{bestEnergy};
    }

    public static void main(String[] args) {
        // Example: Minimize a simple objective function using Simulated Annealing
        Function<double[], Double> objectiveFunction = x -> {
            double sum = 0;
            for (double value : x) {
                sum += Math.pow(value, 2);
            }
            return sum;
        };

        double[] initialSolution = new double[]{2.0, 3.0, 1.5};
        double temperature = 100.0;
        double coolingRate = 0.95;
        int maxIter = 1000;

        double[] result = simulatedAnnealing(objectiveFunction, initialSolution, temperature, coolingRate, maxIter);
        System.out.println("Best energy: " + result[0]);
    }
}
